common/persistence/cassandra/metadata_store.go

```diff
@@ -161,22 +168,35 @@ func (m *MetadataStore) CreateNamespaceInV2Table(
 
 	previous := make(map[string]interface{})
 	applied, iter, err := m.session.MapExecuteBatchCAS(batch, previous)
+	deleteOrphanNamespace := func() {
+		// Delete namespace from `namespaces_by_id`
+		if errDelete := m.session.Query(templateDeleteNamespaceQuery, request.ID).WithContext(ctx).Exec(); errDelete != nil {
+			m.logger.Warn("Unable to delete orphan namespace record. Error", tag.Error(errDelete))
+		}
+	}
+
 	if err != nil {
 		return nil, serviceerror.NewUnavailable(fmt.Sprintf("CreateNamespace operation failed. Inserting into namespaces table. Error: %v", err))
 	}
+
 	defer func() { _ = iter.Close() }()
 
 	if !applied {
-		// Namespace already exist.  Delete orphan namespace record before returning back to user
-		if errDelete := m.session.Query(templateDeleteNamespaceQuery, request.ID).WithContext(ctx).Exec(); errDelete != nil {
-			m.logger.Warn("Unable to delete orphan namespace record. Error", tag.Error(errDelete))
-		}
+		if id, ok := previous["id"]; ok {
+			id = gocql.UUIDToString(id)
+			if id != request.ID {
+				// Namespace with the given name already exists with a different ID.
+				// Delete orphan namespace record before returning back to user
+				deleteOrphanNamespace()
+			}
 
-		if id, ok := previous["Id"].([]byte); ok {
-			msg := fmt.Sprintf("Namespace already exists.  NamespaceId: %v", primitives.UUIDString(id))
+			msg := fmt.Sprintf("Namespace already exists.  NamespaceId: %v", id)
 			return nil, serviceerror.NewNamespaceAlreadyExists(msg)
 		}
 
+		// If namespace does not exist already and it failed to insert, there is a conditional failure.
+		// Delete orphan namespace record before returning back to user
+		deleteOrphanNamespace()
 		return nil, serviceerror.NewNamespaceAlreadyExists("CreateNamespace operation failed because of conditional failure.")
```

Comment:
It looks weird to return NamespaceAlreadyExists error, which is non-retryable, when error is actually caused by conditional update failure, due to for example concurrent namespace registration/update.

Reply:
I agree, how about returning `Unavailable` error?

Reply:
👍 sounds good to me.